lib/prawn/markup/processor/blocks.rb

```diff
@@ -63,10 +63,15 @@ def append_new_line
       def add_paragraph
         text = dump_text
         text.gsub!(/[^\n]/, '') if text.strip.empty?
-        unless text.empty?
+
+        if !text.empty?
           add_bottom_margin
           add_formatted_text(text, text_options)
           put_bottom_margin(text_margin_bottom)
+        elsif text.empty? && treat_empty_paragraph_as_new_line?
+          add_bottom_margin
+          append_text("\n")
+          put_bottom_margin(text_margin_bottom)
```

Comment:
I do not think that `append_text` is the correct thing to do here, because that adds the string to the buffer but not to the document (or only in a subsequent call). Would the following work as well?

```ruby
...
return if text.empty? && !treat_empty_paragraph_as_new_line?

add_bottom_margin
add_formatted_text(text, text_options)
put_bottom_margin(text_margin_bottom)
```

Reply:
Does `add_formatted_text("\n")` do anything? So we could call `add_formatted_text(text.empty? ? "\n" : text, text_options)`.